### PR TITLE
`Assessment`: Add consistent handling if no more unassessed submissions are available

### PR DESCRIPTION
**Pull request by simon-hng**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Description
The endpoints for getting submission without assessments
- `/exercises/{exerciseId}/modeling-submission-without-assessment`
- `/exercises/{exerciseId}/programming-submission-without-assessment`
- `/exercises/{exerciseId}/text-submission-without-assessment`
- `/exercises/{exerciseId}/file-upload-submission-without-assessment`

used to throw an exception that returns an 404 error to the client.
Since not having anymore unassessed submissions is normal behavior, returning null is a better solution. 

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student

1. instructor: Setup exercises for each type i.e. Programming Exercise, Modeling Exercise, etc.
2. student: Create submissions for each exercise
3. instructor: Check if you can assess/access the submissions for the exercise dashboards.

(Additionally check if you get an error after reaching the last submission)

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java [hunk 1]
```diff
@@ -154,27 +153,31 @@ else if (modelingExercise.isExamExercise()) {
...
-        var submissionWithoutResult = super.getRandomAssessableSubmission(modelingExercise, isExamMode, correctionRound)
-                .orElseThrow(() -> new EntityNotFoundException("Modeling submission for exercise " + modelingExercise.getId() + " could not be found"));
-        ModelingSubmission modelingSubmission = (ModelingSubmission) submissionWithoutResult;
+    public Optional<ModelingSubmission> findRandomSubmissionWithoutExistingAssessment(boolean lockSubmission, int correctionRound, ModelingExercise modelingExercise,
+            boolean isExamMode) {
+        var submissionWithoutResult = super.getRandomAssessableSubmission(modelingExercise, isExamMode, correctionRound);
+        if (submissionWithoutResult.isEmpty())
+            return Optional.empty();
+
+        ModelingSubmission modelingSubmission = (ModelingSubmission) submissionWithoutResult.get();
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java:171]: This cast seems unnecessary. 
```suggestion
        ModelingSubmission modelingSubmission = submissionWithoutResult.get();
```

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts [hunk 2]
```diff
@@ -570,6 +570,7 @@ export class ExerciseAssessmentDashboardComponent implements OnInit {
                     if (this.unassessedSubmissionByRound) {
                         this.unassessedSubmissionByRound.delete(correctionRound);
                     }
+                    return;
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:573]: Why is this needed?

File: src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java [hunk 3]
```diff
@@ -589,7 +589,8 @@ void getModelSubmissionWithoutAssessment_noSubmittedSubmission_notFound() throws
         database.addModelingSubmission(classExercise, submission, "student1");
         database.updateExerciseDueDate(classExercise.getId(), ZonedDateTime.now().minusHours(1));
 
-        request.get("/api/exercises/" + classExercise.getId() + "/modeling-submission-without-assessment", HttpStatus.NOT_FOUND, ModelingSubmission.class);
+        var response = request.get("/api/exercises/" + classExercise.getId() + "/modeling-submission-without-assessment", HttpStatus.OK, ModelingSubmission.class);
```
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java:592]: Also change the method name for this test (name contains "not found"), same below.

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts (hunk 2 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:573]: Was actually a refactoring error by myself since before it was handled as an `error:`, meaning that the other code in `next:` didn't run

File: src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java (hunk 1 above)
Comment (simon-hng) [reply src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java:171]: `super.getRandomAssessableSubmission` returns a `Submission`, which is the super class of `ModelingSubmission`. With the method returning a `ModelingSubmission`, I might as well just cast it there.

File: src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java (hunk 3 above)
Comment (simon-hng) [reply src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java:592]: resolved in https://github.com/ls1intum/Artemis/pull/5795/commits/2e9e90205be8c981b9391f6d7a1cda741952a8b5

Comment (Strohgelaender) [reply src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java:592]: Same for `TextSubmissionIntegrationTest.java`

File: src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java [hunk 4]
```diff
@@ -154,27 +153,31 @@ else if (modelingExercise.isExamExercise()) {
...
      */
-    public ModelingSubmission findRandomSubmissionWithoutExistingAssessment(boolean lockSubmission, int correctionRound, ModelingExercise modelingExercise, boolean isExamMode) {
-        var submissionWithoutResult = super.getRandomAssessableSubmission(modelingExercise, isExamMode, correctionRound)
-                .orElseThrow(() -> new EntityNotFoundException("Modeling submission for exercise " + modelingExercise.getId() + " could not be found"));
-        ModelingSubmission modelingSubmission = (ModelingSubmission) submissionWithoutResult;
+    public Optional<ModelingSubmission> findRandomSubmissionWithoutExistingAssessment(boolean lockSubmission, int correctionRound, ModelingExercise modelingExercise,
+            boolean isExamMode) {
+        var submissionWithoutResult = super.getRandomAssessableSubmission(modelingExercise, isExamMode, correctionRound);
+        if (submissionWithoutResult.isEmpty())
+            return Optional.empty();
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java:169]:
```suggestion
        if (submissionWithoutResult.isEmpty()) {
            return Optional.empty();
        }
```

File: src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java (hunk 3 above)
Comment (simon-hng) [reply src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java:592]: done

**Review by Strohgelaender — DISMISSED** [review]
code look good.

**Review by Santia-go — COMMENTED** [review]
Tested on TS1
I am not sure if your PR is related to this, but when I reach the last submission I get this

https://user-images.githubusercontent.com/88991954/199512314-e5a9a65b-aa01-429f-aa63-8b25644eaf2c.mov

File: src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts [hunk 5]
```diff
@@ -281,13 +290,8 @@ export class FileUploadAssessmentComponent implements OnInit, OnDestroy {
...
             },
             error: (error: HttpErrorResponse) => {
-                if (error.status === 404) {
-                    // there are no unassessed submission, nothing we have to worry about
-                    this.hasNewSubmissions = false;
-                } else {
-                    this.isLoading = false;
-                    onError(this.alertService, error);
-                }
+                this.isLoading = false;
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts:293]: Could be moved into a finalize operator

**Review by stefanwaldhauser — APPROVED** [review]
Looks good! GJ :)

File: src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts (hunk 5 above)
Comment (simon-hng) [reply src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts:293]: nice suggestion. I will look into it

**Review by manuelmanso — APPROVED** [review]
Tested in ts2, works as expected. Code looks good too!

**Review by akesfeden — APPROVED** [review]
Tested in testing session on TS2, no error message is displayed when there are no more exercises to assess.

**Review by Strohgelaender — APPROVED** [review]

**Review by kekskatze — CHANGES_REQUESTED** [review]
After assessing the last submission if you click on "Assess next solution" on a modeling exercise nothing happened. It doesn't throw an error (like the testing instructions specify), but it also doesn't open the "We couldn't find any unassessed submissions."-page (on the text exercise it works)

**Comment by simon-hng** [comment]
@kekskatze good catch. I noted that and will look into further inconsistencies in the follow up
